File: samples/dotnet-aspnet/src/Safe/SafeReportRepository.cs
using Microsoft.Data.SqlClient;

namespace Reporting.Safe;

/// <summary>
/// Correct counterpart of ReportRepository.
///
/// Same provider, same query, same names.
/// The owner is bound as a parameter rather than concatenated into the statement.
///
/// Any finding reported in this file is a false positive.
/// </summary>
public class SafeReportRepository
{
    private readonly SafeCryptoService _crypto;

    public SafeReportRepository(SafeCryptoService crypto) => _crypto = crypto;

    public async Task<List<string>> FindByOwnerAsync(string owner)
    {
        var titles = new List<string>();

        await using var connection = new SqlConnection(_crypto.GetConnectionString());
        await connection.OpenAsync();

        const string sql = "SELECT title FROM reports WHERE owner = @owner";
        await using var command = new SqlCommand(sql, connection);
        command.Parameters.AddWithValue("@owner", owner);

        await using var reader = await command.ExecuteReaderAsync();
        while (await reader.ReadAsync())
        {
            titles.Add(reader.GetString(0));
        }

        return titles;
    }
}

File: samples/java-spring/src/main/java/com/example/reporting/safe/SafeReportRepository.java
package com.example.reporting.safe;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.List;
import javax.sql.DataSource;
import org.springframework.stereotype.Repository;

/**
 * Correct counterpart of ReportRepository.
 *
 * Same driver, same query, same names.
 * The owner is bound as a parameter rather than concatenated into the statement.
 *
 * Any finding reported in this file is a false positive.
 */
@Repository
public class SafeReportRepository {

    private final DataSource dataSource;

    public SafeReportRepository(DataSource dataSource) {
        this.dataSource = dataSource;
    }

    public List<String> findByOwner(String owner) throws Exception {
        List<String> titles = new ArrayList<>();

        String sql = "SELECT title FROM reports WHERE owner = ?";

        try (Connection connection = dataSource.getConnection();
                PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setString(1, owner);
            ResultSet results = statement.executeQuery();

            while (results.next()) {
                titles.add(results.getString("title"));
            }
        }

        return titles;
    }
}
